crates/sui-types/src/move_package.rs

```diff
@@ -320,3 +433,104 @@ where
     }
     normalized_modules
 }
+
+fn build_linkage_table<'p>(
+    mut immediate_dependencies: BTreeSet<ObjectID>,
+    transitive_dependencies: impl IntoIterator<Item = &'p MovePackage>,
+) -> Result<BTreeMap<ObjectID, UpgradeInfo>, ExecutionError> {
+    let mut linkage_table = BTreeMap::new();
+    let mut dep_linkage_tables = vec![];
+
+    for transitive_dep in transitive_dependencies.into_iter() {
+        let original_id = transitive_dep.original_package_id();
+
+        if immediate_dependencies.remove(&original_id) {
+            // Found an immediate dependency, mark it as seen, and stash a reference to its linkage
+            // table to check later.
+            dep_linkage_tables.push(&transitive_dep.linkage_table);
+        }
+
+        linkage_table.insert(
+            original_id,
+            UpgradeInfo {
+                upgraded_id: transitive_dep.id,
+                upgraded_version: transitive_dep.version,
+            },
+        );
+    }
+    // (1) Every dependency is represented in the transitive dependencies
+    if !immediate_dependencies.is_empty() {
+        // TODO ExecutionError
```

Comment:
Flagging TODO -- for all the `ExecutionError` ones, I think @awelc, you may have had some code already to hand (in your divergent version of this branch) to handle this.